Readme.md
# MQTT2AWSS3Bridge

MQTT2AWSS3Bridge is a project to connect a locally running broker to AWS S3 and upload the messages as a simple .txt file. The project was written and tested in Python 3.

[![Build status](https://ci.appveyor.com/api/projects/status/q54cv976lrsclueu?svg=true)](https://ci.appveyor.com/project/SeppPenner/mqtt2awss3bridge)
[![GitHub issues](https://img.shields.io/github/issues/SeppPenner/MQTT2AWSS3Bridge.svg)](https://github.com/SeppPenner/MQTT2AWSS3Bridge/issues)
[![GitHub forks](https://img.shields.io/github/forks/SeppPenner/MQTT2AWSS3Bridge.svg)](https://github.com/SeppPenner/MQTT2AWSS3Bridge/network)
[![GitHub stars](https://img.shields.io/github/stars/SeppPenner/MQTT2AWSS3Bridge.svg)](https://github.com/SeppPenner/MQTT2AWSS3Bridge/stargazers)
[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://raw.githubusercontent.com/SeppPenner/MQTT2AWSS3Bridge/master/License.txt)
[![Known Vulnerabilities](https://snyk.io/test/github/SeppPenner/MQTT2AWSS3Bridge/badge.svg)](https://snyk.io/test/github/SeppPenner/MQTT2AWSS3Bridge) 

## Adjust your settings:

* Adjust the broker to the address you want to use: `broker_source`
* Add your custom filters to `filterMessage()` if you want to filter messages
* Adjust your credentials (uncomment if anonymous): 

```python
client_source.username_pw_set("mqtt", "IoT")
```

* Install the AWS command line: https://docs.aws.amazon.com/cli/latest/userguide/installing.html

* Adjust your AWS S3 credentials and the proxy settings:

```python
os.environ['HTTPS_PROXY'] = 'http://YourProxyIfNeeded:8080'
os.environ['AWS_ACCESS_KEY_ID'] = 'YourKeyHere'
os.environ['AWS_SECRET_ACCESS_KEY'] = 'YourSecretHere'
os.environ['AWS_DEFAULT_REGION'] = 'eu-central-1'
os.environ['AWS_SESSION_TOKEN'] ='YourSessionTokenHere'
yourAWSBucket = 'YourAWSBucketHere'
```

* Add filters to the bridging like described in the `bridgeFiltering.py` file if needed:

```python
def filterMessage(payload, topic, qos):
	"Filters the messages depending on the configuration for the attributes payload, topic and QoS. 'True' means that the message is not forwarded."
	# Examples below:
	if(payload == "10 %"):
		print('Filtered: payload == "10 %"')
		return True
	if(topic == "humidity" and qos == 0):
		print('Filtered: topic == "humidity" and qos == 0')
		return True
	if(topic == "temperature" or qos == 2):
		print('Filtered: topic == "temperature" or qos == 2')
		return True
	#Add your filters here
```

## Setup on the Raspberry Pi

```bash
sudo apt-get install python3
sudo apt-get install python3-pip
sudo pip3 install paho-mqtt
sudo pip3 install awscli --upgrade --user
```

or

```bash
sudo apt-get install python3
sudo apt-get install python3-pip
sudo pip3 install -r requirements.txt
```

## Running the programms:

```bash
python3 bridge.py
python3 bridgeFiltering.py
```

## Installing the latest version of Python (Currently 3.7.4) on the Raspberry Pi:

https://gist.github.com/SeppPenner/6a5a30ebc8f79936fa136c524417761d

## Paho MQTT client documentation

* https://pypi.org/project/paho-mqtt/
* https://www.hivemq.com/blog/mqtt-client-library-paho-python

## See also

* [MQTT2MQTTBridge](https://github.com/SeppPenner/MQTT2MQTTBridge)
* [MQTT2PostgresBridge](https://github.com/SeppPenner/MQTT2PostgresBridge)
* [MQTT2MySQLBridge](https://github.com/SeppPenner/MQTT2MySQLBridge)

Change history
--------------

See the [Changelog](https://github.com/SeppPenner/MQTT2AWSS3Bridge/blob/master/Changelog.md).

Changelog.md
Change history
--------------

* **Version 1.0.3.0 (2021-07-25)** : Updated requirements.
* **Version 1.0.2.0 (2020-06-12)** : Updated AWSCLI package.
* **Version 1.0.1.0 (2020-06-04)** : Removed license year, added first tag.
* **Version 1.0.0.1 (2019-09-29)** : Updated python version, updated requirements.
* **Version 1.0.0.0 (?)** : 1.0 release.

bridge.py
import paho.mqtt.client as mqtt
import os
import subprocess
import sys
import datetime
import traceback

broker_source = "127.0.0.1"
broker_source_port = 1883
client_source = mqtt.Client("YourClientId")
client_source.username_pw_set("YourUser", "YourPassword")

os.environ['HTTPS_PROXY'] = 'http://YourProxyIfNeeded:8080'
os.environ['AWS_ACCESS_KEY_ID'] = 'YourKeyHere'
os.environ['AWS_SECRET_ACCESS_KEY'] = 'YourSecretHere'
os.environ['AWS_DEFAULT_REGION'] = 'eu-central-1'
os.environ['AWS_SESSION_TOKEN'] ='YourSessionTokenHere'
yourAWSBucket = 'YourAWSBucketHere'

resultsFile = ""

def on_message(client, userdata, message):
	"Evaluated when a new message is received on a subscribed topic"
	print("Received message '" + str(message.payload)[2:][:-1] + "' on topic '"
		+ message.topic + "' with QoS " + str(message.qos))
	writeToResultsFile(message)

def writeToResultsFile(message):
	"Writes to the result file or creates a new one if needed"
	print("Resultsfile is " + resultsFile)
	if(os.stat(resultsFile).st_size >= 50000000): #Splits the file at 50 MB size
		copyFileToS3(resultsFile)
		createNewFile()
	with open(resultsFile, 'a') as f:
		f.write('"' + str(message.payload)[2:][:-1] + '","' + str(message.topic) + '","' + str(message.qos) + '"\n')

def createNewFile():
	"Creates a new results file"
	global resultsFile
	resultsFile = getCurrentDateTimeFormatted() + '.txt'
	with open(resultsFile, 'w') as f:
		f.write('"Payload", "Topic", "QoS"\n')
	print("New file " + resultsFile + " created")

def setup():
	"Runs the setup procedure for the client"
	print("Setting up the onMessage handler")
	client_source.on_message = on_message
	print("Connecting to source")
	client_source.connect(broker_source, broker_source_port)
	client_source.subscribe("#", qos=1)
	print("Setup finished, waiting for messages...")

def copyFileToS3(fileName):
	"Copies the given file to the given Amazon S3 bucket"
	print("Copying file to " + yourAWSBucket + str(fileName.replace("\\", "/")))
	subprocess.call(["aws", "s3", "cp", str(fileName), yourAWSBucket + str(fileName.replace("\\", "/"))])

def getCurrentDateTimeFormatted():
	"Returns a string for the output file with the current date and time"
	return datetime.datetime.now().isoformat().replace(":", "-").replace(".", "-")

try:
	createNewFile()
	setup()
	client_source.loop_forever()
except Exception as e:
	traceback.print_exc()

bridgeS3.py
import paho.mqtt.client as mqtt
import os
import subprocess
import sys
import datetime
import traceback

broker_source = "127.0.0.1"
broker_source_port = 1883
client_source = mqtt.Client("YourClientId")
client_source.username_pw_set("YourUser", "YourPassword")

os.environ['HTTPS_PROXY'] = 'http://YourProxyIfNeeded:8080'
os.environ['AWS_ACCESS_KEY_ID'] = 'YourKeyHere'
os.environ['AWS_SECRET_ACCESS_KEY'] = 'YourSecretHere'
os.environ['AWS_DEFAULT_REGION'] = 'eu-central-1'
os.environ['AWS_SESSION_TOKEN'] ='YourSessionTokenHere'
yourAWSBucket = 'YourAWSBucketHere'

resultsFile = ""

def on_message(client, userdata, message):
	"Evaluated when a new message is received on a subscribed topic"
	print("Received message '" + str(message.payload) + "' on topic '"
		+ message.topic + "' with QoS " + str(message.qos))
	writeToResultsFile(message)

def writeToResultsFile(message):
	"Writes to the result file or creates a new one if needed"
	print("Resultsfile is " + resultsFile)
	if(os.stat(resultsFile).st_size >= 50000000): #Splits the file at 50 MB size
		copyFileToS3(resultsFile)
		createNewFile()
	with open(resultsFile, 'a') as f:
		f.write('"' + str(message.payload).replace("b'", "").replace("'", "") + '","' + str(message.topic) + '","' + str(message.qos) + '"\n')

def createNewFile():
	"Creates a new results file"
	global resultsFile
	resultsFile = getCurrentDateTimeFormatted() + '.txt'
	with open(resultsFile, 'w') as f:
		f.write('"Payload", "Topic", "QoS"\n')
	print("New file " + resultsFile + " created")

def setup():
	"Runs the setup procedure for the client"
	print("Setting up the onMessage handler")
	client_source.on_message = on_message
	print("Connecting to source")
	client_source.connect(broker_source, broker_source_port)
	client_source.subscribe("#", qos=1)
	print("Setup finished, waiting for messages...")

def copyFileToS3(fileName):
	"Copies the given file to the given Amazon S3 bucket"
	print("Copying file to " + yourAWSBucket + str(fileName.replace("\\", "/")))
	subprocess.call(["aws", "s3", "cp", str(fileName), yourAWSBucket + str(fileName.replace("\\", "/"))])

def getCurrentDateTimeFormatted():
	"Returns a string for the output file with the current date and time"
	return datetime.datetime.now().isoformat().replace(":", "-").replace(".", "-")

try:
	createNewFile()
	setup()
	client_source.loop_forever()
except Exception as e:
	traceback.print_exc()

bridgeFiltering.py
import paho.mqtt.client as mqtt
import os
import subprocess
import sys
import datetime
import traceback

broker_source = "127.0.0.1"
broker_source_port = 1883
client_source = mqtt.Client("YourClientId")
client_source.username_pw_set("YourUser", "YourPassword")

os.environ['HTTPS_PROXY'] = 'http://YourProxyIfNeeded:8080'
os.environ['AWS_ACCESS_KEY_ID'] = 'YourKeyHere'
os.environ['AWS_SECRET_ACCESS_KEY'] = 'YourSecretHere'
os.environ['AWS_DEFAULT_REGION'] = 'eu-central-1'
os.environ['AWS_SESSION_TOKEN'] ='YourSessionTokenHere'
yourAWSBucket = 'YourAWSBucketHere'

resultsFile = ""

def on_message(client, userdata, message):
	"Evaluated when a new message is received on a subscribed topic"
	print("Received message '" + str(message.payload)[2:][:-1] + "' on topic '"
		+ message.topic + "' with QoS " + str(message.qos))
	if not filterMessage(str(message.payload)[2:][:-1], str(message.topic), (message.qos)):
		writeToResultsFile(message)

def writeToResultsFile(message):
	"Writes to the result file or creates a new one if needed"
	print("Resultsfile is " + resultsFile)
	if(os.stat(resultsFile).st_size >= 50000000): #Splits the file at 50 MB size
		copyFileToS3(resultsFile)
		createNewFile()
	with open(resultsFile, 'a') as f:
		f.write('"' + str(message.payload)[2:][:-1] + '","' + str(message.topic) + '","' + str(message.qos) + '"\n')

def createNewFile():
	"Creates a new results file"
	global resultsFile
	resultsFile = getCurrentDateTimeFormatted() + '.txt'
	with open(resultsFile, 'w') as f:
		f.write('"Payload", "Topic", "QoS"\n')
	print("New file " + resultsFile + " created")

def setup():
	"Runs the setup procedure for the client"
	print("Setting up the onMessage handler")
	client_source.on_message = on_message
	print("Connecting to source")
	client_source.connect(broker_source, broker_source_port)
	client_source.subscribe("#", qos=1)
	print("Setup finished, waiting for messages...")
	
def filterMessage(payload, topic, qos):
	"Filters the messages depending on the configuration for the attributes payload, topic and QoS. 'True' means that the message is not forwarded."
	# Examples below:
	if(payload == "10 %"):
		print('Filtered: payload == "10 %"')
		return True
	if(topic == "humidity" and qos == 0):
		print('Filtered: topic == "humidity" and qos == 0')
		return True
	if(topic == "temperature" or qos == 2):
		print('Filtered: topic == "temperature" or qos == 2')
		return True
	#Add your filters here

def copyFileToS3(fileName):
	"Copies the given file to the given Amazon S3 bucket"
	print("Copying file to " + yourAWSBucket + str(fileName.replace("\\", "/")))
	subprocess.call(["aws", "s3", "cp", str(fileName), yourAWSBucket + str(fileName.replace("\\", "/"))])

def getCurrentDateTimeFormatted():
	"Returns a string for the output file with the current date and time"
	return datetime.datetime.now().isoformat().replace(":", "-").replace(".", "-")

try:
	createNewFile()
	setup()
	client_source.loop_forever()
except Exception as e:
	traceback.print_exc()